crates/holochain_state/src/source_chain.rs

```diff
@@ -300,10 +283,77 @@ impl SourceChain {
         let author = self.author.clone();
         let persisted_head = self.head_info.as_ref().map(|h| h.action.clone());
 
+        let now = Timestamp::now();
+
+        // Take out a write lock as late as possible, after doing everything we can in memory and
+        // before starting any database read/write operations.
+        let write_permit = self.vault.acquire_write_permit().await?;
+
+        self.vault
+            .read_async({
```

Comment:
I think I'm cool with this as-is. But just for consideration: If the write permit is held here anyways, and we're beholden to the chc call in the middle no matter the case, would it make more sense to actually just do all this logic (including the chc call) inside the `write_async` transaction below?

Reply:
It seems wrong to do a network operation inside a database transaction but I do see the logic. I was thinking that returning the connection to the pool would allow it to be used for reads but actually the pool is locally partitioned by the sempahores so that won't happen